### DISCUSSION TITLE
Since trivy:0.49.0 several jar files with qualifier are no longer detected

### DISCUSSION THREAD

**Discussion by bpapez** [aquasecurity/trivy#6233] [Bugs]
### Description

In our project we create the SBOM with the Trivy image scan (format cyclonedy). Since Trivy version 0.49.0 the SBOM is no longer containing several JARs, which use a qualifier other than SNAPSHOT.

### Desired Behavior

Dependencies with custom qualifiers in the version should still be detected and listed in the SBOM.

### Actual Behavior

For instance [jasypt-1.9.3-lite.jar](https://repo1.maven.org/maven2/org/jasypt/jasypt/1.9.3/). The log shows

> No such POM in the central repositories {"file": "jasypt-1.9.3-lite.jar"}

although it exists in a central repository (see link above). It still worked correctly with Trivy 0.48.3. 

We have about 30 dependencies like that, which now disappeared from the SBOM. Some of these jars are built by us from forked repositories, where we either backported a patch or where we had to make a JAR work well with OSGI. For that we keep the same version, but we add a custom qualifier, which is a valid practice. Also Dependency Track was able to report vulnerabilities on them, therefore it is important that such files do not get simply removed from the SBOM. 

For now we have to stick with Trivy 0.48.3 . Could this get fixed again please or is there an option we could use to still get a correct SBOM with the latest Trivy? I tried `--offline-scan`, but that did not make a difference. 

### Reproduction Steps

```bash
1. I am scanning an image.tar with trivy image --format cyclonedx --output dependency-results.sbom.cdx.json --input image.tar
2. The sbom file no longer contains dependency to jasypt-1.9.3-lite.jar and others, although with trivy 0.48.3 they were still listed.
```

### Target

Container Image

### Scanner

None

### Output Format

CycloneDX

### Mode

Standalone

### Debug Output

```bash
2024-02-29T12:29:58.871Z        DEBUG   Parsing Java artifacts...       {"file": "usr/local/tomcat/webapps/ROOT/WEB-INF/lib/jasypt-1.9.3-lite.jar"}
2024-02-29T12:29:58.872Z        DEBUG   No such POM in the central repositories {"file": "jasypt-1.9.3-lite.jar"}
```

### Operating System

I use the provided Docker image on different OS

### Version

```bash
Version: 0.49.1
Java DB:
  Version: 1
  UpdatedAt: 2024-02-29 00:45:37.104370989 +0000 UTC
  NextUpdate: 2024-03-03 00:45:37.104370869 +0000 UTC
  DownloadedAt: 2024-02-29 09:56:46.4002486 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Hello @bpapez 
Thanks for your report!

At the moment I assume the following points:
1) your jar file doesn't contain GAV (GroupID ArtifactID and version) in `pom.properties` and `MANIFEST.MF` files.
2) your jar file is not same as jar file from maven repository (so we can't detect this artifact using sha1)
3) we updated the logic to avoid false positives: now when we try to guess GroupID by filename - we only check artifact if maven repository contains version from filename. If you changed file name (used updated version), maven repository doesn't contain this version and we skip this artifact.

If I'm right, I think best solution for you is to populate `pom.properties` file with new version. In this case, Trivy will be able to correctly detect the artifact (you will also save time and resources, since Trivy does not need to look for artifacts from trivy-java-db).

Regards, Dmitriy

**Reply by bpapez** [reply]
Unfortunately most of these libraries are not ours, but are open source libraries or transitive dependencies from 3rd party libraries we use. They are dowloaded from the maven repository so the sha1 is correct. 

I gave you as example the [jasypt-1.9.3-lite.jar](https://repo1.maven.org/maven2/org/jasypt/jasypt/1.9.3/), but there is also [commons-jexl-1.1.jar](https://repo1.maven.org/maven2/commons-jexl/commons-jexl/1.1), [stax-1.2.0.jar](https://repo1.maven.org/maven2/stax/stax/1.2.0), [xmlpull-1.1.3.1.jar](https://repo1.maven.org/maven2/xmlpull/xmlpull/1.1.3.1), [urlrewritefilter-3.2.0.jar](https://repo1.maven.org/maven2/org/tuckey/urlrewritefilter/3.2.0) to name a few more. They are all downloaded from maven repository, so the sha1 is correct, its just that they do not have a `pom.properties` inside. Hundreds of projects are using these libraries. 

There are also many valid cases where a company changes the filename of an artifact, by not changing the version name but by adding a qualifier in order to backport a fix in a no longer maintained library version or like in our case to make a library OSGI ready. For instance we used `spring-core-3.2.18.RELEASE_OSGI.jar` or we added our company name in the qualifier after the version. We are uploading them just to our own internal Nexus repository. With Trivy 0.48.x these libraries still got detected and listed in the SBOM.

As said about 30 dependencies found in WEB-INF/lib of a single project are no longer listed with Trivy 49.0. All of them were not "false positives". I'm sure that even more valid libraries are getting missed in other folders or projects.

If Trivy no longer detects them although they are there and are not false positives and there is no way to get them listed again in the created SBOM, then Trivy unfortunately no longer is a reliable tool to create SBOMs.

Regards,
Benjamin

**Reply by DmitriyLewen** [reply]
Hello @bpapez 
Thanks for your examples.

I see your issue now (Unfortunately, i missed `lite` suffix in your 1st example at first).

I found reason of this problem:
We parse `metadata.xml` file for artifact and check found versions.
But looks like we can't trust of this file.
Your examples show that we don't include part of artifacts into java-db.
So we need to upgrade our logic for collecting artifact.
I created https://github.com/aquasecurity/trivy-java-db/issues/28 for this task.

**Reply by DmitriyLewen** [reply]
Hello @bpapez 
I created https://github.com/aquasecurity/trivy-java-db/issues/29 to update logic of trivy-java-db.

Can you try using [new DB](https://github.com/DmitriyLewen/trivy-java-db/pkgs/container/trivy-java-db/189767366?tag=1) (you can use `--java-db-repository ghcr.io/dmitriylewen/trivy-java-db` for this), if you have time?

Regards, Dmitriy

**Reply by bpapez** [reply]
Hello Dmitriy,
I tried it today by just adding that parameter in my trivy call, but it does not make any difference. So I am not sure it is actually taking that DB. Do I also have to download it first? I tried to call `docker pull ghcr.io/dmitriylewen/trivy-java-db:1`, but I get 
```
1: Pulling from dmitriylewen/trivy-java-db
unsupported media type application/vnd.oci.empty.v1+json
```

Regards,
Benjamin

**Reply by DmitriyLewen** [reply]
`INFO` log shows repository:
```
2024-04-15T08:54:09.501+0600	INFO	Java DB Repository: ghcr.io/dmitriylewen/trivy-java-db:1
```


___

> unsupported media type application/vnd.oci.empty.v1+json

To download trivy-db or trivy-java-db you need to use `oras` - https://aquasecurity.github.io/trivy/v0.50/docs/advanced/air-gap/#download-the-vulnerability-database

**Reply by knqyf263** [reply]
or Trivy

```
$ trivy --download-db-only
```

**Reply by bpapez** [reply]
I was able to use that new DB now and did the analysis on my test image. From the 37 libraries, which got missed from the SBOM with Trivy 0.49.0 there are 8 libraries, which came back with the DB you provided:
```
"purl": "pkg:maven/commons-jexl/commons-jexl@1.1"  (was: "purl": "pkg:maven/org.jvnet.hudson/commons-jexl@1.1", )
"purl": "pkg:maven/com.google.inject/guice@4.2.2-no_aop" (was: "purl": "pkg:maven/com.guicedee.services/guice@4.2.2-no_aop")
"purl": "pkg:maven/org.jasypt/jasypt@1.9.3-lite"
"purl": "pkg:maven/jaxen/jaxen@1.1.1" (was: "purl": "pkg:maven/com.guicedee.services/jaxen@1.1.1")
"purl": "pkg:maven/stax/stax@1.2.0"
"purl": "pkg:maven/org.tuckey/urlrewritefilter@3.2.0"
"purl": "pkg:maven/org.apache.maven.wagon/wagon-http@3.5.3-shaded"
"purl": "pkg:maven/xmlpull/xmlpull@1.1.3.1"
```
Three of them had a different package, when I ran the scan with Trivy 0.48.3 on the original DB. I think the current package name from your provided DB is better now.

Regarding these package changes there is also one library, where with your DB the package name changed and in my previous runs with 0.48.3 and later 0.49.x it also already changed:
```
"purl": "pkg:maven/antlr/antlr@2.7.7" (was: "purl": "pkg:maven/io.github.jglrxavpok.hephaistos/antlr@2.7.7" 
and later: "purl": "pkg:maven/org.glassfish.external/antlr@2.7.7" )
```
The new purl from your DB is also the best: `pkg:maven/antlr/antlr@2.7.7`

Unfortunately there are still 29 libraries, which are missed when testing with your new DB:
```
      "purl": "pkg:maven/com.proofpoint.platform/concurrent@1.3.4-jahia6"
      "purl": "pkg:maven/com.syncthemall/boilerpipe@1.1.0-jahia1"
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.20"
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.1"
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.3"
      "purl": "pkg:maven/io.github.valters/daisydiff@1.1-nodeps"
      "purl": "pkg:maven/org.apache.jackrabbit/jackrabbit-data@2.20.12-jahia1-SNAPSHOT-tests"
      "purl": "pkg:maven/org.apache.jackrabbit/jackrabbit-webapp@2.20.12-jahia1-SNAPSHOT"
      "purl": "pkg:maven/org.apache.solr/solr-core@3.6.2-jahia1"
      "purl": "pkg:maven/org.apache.solr/solr-solrj@3.6.2-jahia1"
      "purl": "pkg:maven/org.drools/drools-compiler@6.0.0.Final-nocdi"
      "purl": "pkg:maven/org.jbpm/jbpm-executor@6.0.0.Final-nocdi"
      "purl": "pkg:maven/org.jbpm/jbpm-human-task-audit@6.0.0.Final-nocdi"
      "purl": "pkg:maven/org.jbpm/jbpm-human-task-core@6.0.0.Final-jahia-nocdi"
      "purl": "pkg:maven/org.jbpm/jbpm-human-task-workitems@6.0.0.Final-nocdi"
      "purl": "pkg:maven/org.jbpm/jbpm-runtime-manager@6.0.0.Final-nocdi"
      "purl": "pkg:maven/org.jbpm/jbpm-shared-services@6.0.0.Final-jahia-nocdi"
      "purl": "pkg:maven/org.springframework/spring-aop@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-beans@3.2.18.jahia1_OSGI"
      "purl": "pkg:maven/org.springframework/spring-context-support@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-context@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-core@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-expression@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-jdbc@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-orm@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-tx@3.2.18.RELEASE_OSGI2"
      "purl": "pkg:maven/org.springframework/spring-web@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/org.springframework/spring-webmvc@3.2.18.RELEASE_OSGI"
      "purl": "pkg:maven/se.l4.commons/commons-id@0.1-dev"
```

Do you have hints why they are missed and what you or we have to do to bring them back? Most of them are artifacts, which have been forked from the original repository and then built by us. Interestingly other such forked and built artifacts are still found and listed by Trivy with your new DB, like:

```
      "purl": "pkg:maven/org.apache.jackrabbit/jackrabbit-core@2.20.12-jahia1-SNAPSHOT"
      "purl": "pkg:maven/org.jbpm/jbpm-human-task-core@6.0.0.Final-jahia"
      "purl": "pkg:maven/org.ops4j.pax.web/pax-web-jsp@7.3.28-jahia1"
      ...
```
Why are these found and the others above are not? 
Also notice that these four are not forked and built by us, but got missed since 0.48.3 even with your new DB:
```
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.20"
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.1"
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.3"
      "purl": "pkg:maven/se.l4.commons/commons-id@0.1-dev"
```

Best regards,
Benjamin

**Reply by DmitriyLewen** [reply]
> Unfortunately there are still 29 libraries, which are missed when testing with your new DB:

I see versions of these packages and they look very unusual. Are you sure maven Central contains these versions?
for example `com.proofpoint.platform/concurrent@1.3.4-jahia6` does not seem to have version 1.3.4 - https://repo1.maven.org/maven2/com/proofpoint/platform/concurrent/

>  built artifacts are still found and listed by Trivy with your new DB, like:
      "purl": "pkg:maven/org.apache.jackrabbit/jackrabbit-core@2.20.12-jahia1-SNAPSHOT"
      "purl": "pkg:maven/org.jbpm/jbpm-human-task-core@6.0.0.Final-jahia"
      "purl": "pkg:maven/org.ops4j.pax.web/pax-web-jsp@7.3.28-jahia1"

Can you share some of these jar? 
The following information will be enough for me:
- jar file name
- pom.properties files
- MANIFEST.MF files

> Also notice that these four are not forked and built by us, but got missed since 0.48.3 even with your new DB:
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.20"
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.1"
      "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.3"

Maven central doesn't have these versions - https://repo1.maven.org/maven2/edu/gmu/cs/ecj/

**Reply by bpapez** [reply]
> for example com.proofpoint.platform/concurrent@1.3.4-jahia6 does not seem to have version 1.3.4 

Trivy does not take the right package/groupId, in reality it is this artifact: https://mvnrepository.com/artifact/concurrent/concurrent and it has version 1.3.4 on central Maven. On the mvnrepository.com page I am linking you can see next to the `Central` tab, also the `Atlassian 3rd-P Old` and then the `Jahia` tab, which is our public Maven repo hosting the `1.3.4-jahia6` version. In the other tabs you can also see that Atlassian, Redhat and others have their own forks of this library.  

We cannot get rid of this old library, because it is a transitive dependency of Apache Jackrabbit, on which we strongly depend. Therefore it would be great if Trivy would detect the right package/groupId and continue listing it in the SBOM. 

> > built artifacts are still found and listed by Trivy with your new DB, like:
> >     "purl": "pkg:maven/org.apache.jackrabbit/[jackrabbit-core@2.20.12-jahia1-SNAPSHOT](mailto:jackrabbit-core@2.20.12-jahia1-SNAPSHOT)"
> >     "purl": "pkg:maven/org.jbpm/[jbpm-human-task-core@6.0.0.Final-jahia](mailto:jbpm-human-task-core@6.0.0.Final-jahia)"
> >     "purl": "pkg:maven/org.ops4j.pax.web/pax-web-jsp@7.3.28-jahia1"
> 
> Can you share some of these jar?
> The following information will be enough for me:
> 
>     jar file name
>     pom.properties files
>     MANIFEST.MF files

They are all on our public repo: https://maven.jahia.org/maven2/
Here are the direct links to the folders with the jar artifacts containing the files you are asking for:
* `jackrabbit-core@2.20.12-jahia1-SNAPSHOT`
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/apache/jackrabbit/jackrabbit-core/2.20.12-jahia1-SNAPSHOT/
* `jbpm-human-task-core@6.0.0.Final-jahia`
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/jbpm/jbpm-human-task-core/6.0.0.Final-jahia/
* `pax-web-jsp@7.3.28-jahia1`
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/ops4j/pax/web/7.3.28-jahia1/

Please notice again that the above three are still correctly found with latest Trivy and the DB you sent me.

But here are links to three artifacts, which are NOT found with latest Trivy and your DB (but they were listed with Trivy 0.48.x)
* `drools-compiler@6.0.0.Final-nocdi`
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/drools/drools-compiler/6.0.0.Final/
* `solr-core@3.6.2-jahia1`
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/apache/solr/solr-core/3.6.2-jahia1/
* `org.springframework/spring-core@3.2.18.RELEASE_OSGI`
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/springframework/spring-core/3.2.18.RELEASE_OSGI/

What should be done to again correctly detect these artifacts and list them in the SBOM?


> >     Also notice that these four are not forked and built by us, but got missed since 0.48.3 even with your new DB:
> >     "purl": "pkg:maven/edu.gmu.cs/ecj@4.20"
> >     "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.1"
> >     "purl": "pkg:maven/edu.gmu.cs/ecj@4.6.3"
> 
> Maven central doesn't have these versions - https://repo1.maven.org/maven2/edu/gmu/cs/ecj/

Also for that Trivy is assuming the wrong package/groupId. The correct groupId is `org.eclipse.jdt.core.compiler`  and Maven central at least has version 4.6.1 : https://repo1.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/

Similarly Trivy is assuming the wrong package/groupId also for:
* `pkg:maven/com.syncthemall/boilerpipe@1.1.0-jahia1` 
which should be `pkg:maven/de.l3s.boilerpipe/boilerpipe@1.1.0-jahia1` [(link to forked artifact)](https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/de/l3s/boilerpipe/boilerpipe/1.1.0-jahia1/) [(link to original on Maven central)](https://repo1.maven.org/maven2/de/l3s/boilerpipe/boilerpipe/1.1.0/)
* `pkg:maven/io.github.valters/daisydiff@1.1-nodeps`
which should be `pkg:maven/org.outerj.daisy/daisydiff@1.1-nodeps` [(link to forked artifact)](https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/outerj/daisy/daisydiff/1.1/) [(other Maven repos have forks)](https://mvnrepository.com/artifact/org.outerj.daisy/daisydiff)
* `pkg:maven/se.l4.commons/commons-id@0.1-dev`
which should be `pkg:maven/commons-id/commons-id@0.1-dev` [(link to forked artifact)](https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/commons-id/commons-id/0.1-dev/)

**Reply by DmitriyLewen** [reply]
> Trivy does not take the right package/groupId, in reality it is this artifact: https://mvnrepository.com/artifact/concurrent/concurrent and it has version 1.3.4 on central Maven. On the mvnrepository.com page I am linking you can see next to the Central tab, also the Atlassian 3rd-P Old and then the Jahia tab, which is our public Maven repo hosting the 1.3.4-jahia6 version. In the other tabs you can also see that Atlassian, Redhat and others have their own forks of this library.

Unfortunately, trivy-java-db only contains maven Central.
trivy-java-db is already large in size, if we add [all repositories](https://mvnrepository.com/repos) size will be very large.

I very want to help you, but it looks like you have only one solution - add information about GAV to these jars and overwrite them..

___

> jackrabbit-core@2.20.12-jahia1-SNAPSHOT
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/apache/jackrabbit/jackrabbit-core/2.20.12-jahia1-SNAPSHOT/

This `jar` file contains correct `pom.properties` file:
```bash
➜ cat ./jackrabbit-core/META-INF/maven/org.apache.jackrabbit/jackrabbit-core/pom.properties 
artifactId=jackrabbit-core
groupId=org.apache.jackrabbit
version=2.20.12-jahia1-SNAPSHOT
```

> But here are links to three artifacts, which are NOT found with latest Trivy and your DB (but they were listed with Trivy 0.48.x)
    drools-compiler@6.0.0.Final-nocdi
    https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/drools/drools-compiler/6.0.0.Final/

It works for me. Can you compare your jar with with file from repository?
```bash
➜ wget https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/drools/drools-compiler/6.0.0.Final/drools-compiler-6.0.0.Final-nocdi.jar
...

drools-compiler-6.0 100%[===================>]   1,44M  1,59MB/s    in 0,9s    

2024-04-24 10:46:38 (1,59 MB/s) - ‘drools-compiler-6.0.0.Final-nocdi.jar’ saved [1507761/1507761]

➜ trivy -q rootfs -f json --list-all-pkgs ./drools-compiler-6.0.0.Final-nocdi.jar | jq ' .Results[].Packages[] | select(.Name=="org.drools:drools-compiler")'
{
  "Name": "org.drools:drools-compiler",
  "Identifier": {
    "PURL": "pkg:maven/org.drools/drools-compiler@6.0.0.Final"
  },
  "Version": "6.0.0.Final",
  "Layer": {},
  "FilePath": "drools-compiler-6.0.0.Final-nocdi.jar"
}
```

____

> Also for that Trivy is assuming the wrong package/groupId. The correct groupId is org.eclipse.jdt.core.compiler and Maven central at least has version 4.6.1 : https://repo1.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/

Can you share link on this jar? Perhaps `pom.properties` or `MANIFEST.MF` file contain `edu.gmu.cs` GroupID.

___ 
> Similarly Trivy is assuming the wrong package/groupId also for:
    pkg:maven/com.syncthemall/boilerpipe@1.1.0-jahia1
    which should be pkg:maven/de.l3s.boilerpipe/boilerpipe@1.1.0-jahia1 [(link to forked artifact)](https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/de/l3s/boilerpipe/boilerpipe/1.1.0-jahia1/) [(link to original on Maven central)](https://repo1.maven.org/maven2/de/l3s/boilerpipe/boilerpipe/1.1.0/)

New Trivy should not detect this can. Are you using the new Trivy for scanning?

**Reply by bpapez** [reply]
>> Trivy does not take the right package/groupId, in reality it is this artifact: https://mvnrepository.com/artifact/concurrent/concurrent and it has version 1.3.4 on central Maven ...

> I very want to help you, but it looks like you have only one solution - add information about GAV to these jars and overwrite them..

I will create internal tickets to try doing that. For the jars, which were created by us from a fork it should not be a problem. 

However I think its bad that we will need to create new forks from 3rd party repos, which we have not changed, just to add a `pom.properties`, which is missing in the original and Trivy's is not able to detect the right package name from `MANIFEST.MF`. 

>> But here are links to three artifacts, which are NOT found with latest Trivy and your DB (but they were listed with Trivy 0.48.x)
[drools-compiler@6.0.0.Final-nocdi](mailto:drools-compiler@6.0.0.Final-nocdi)
https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/drools/drools-compiler/6.0.0.Final/

> It works for me. Can you compare your jar with with file from repository?

Sorry, my mistake. That jar is getting correctly detected. I shouldn't have placed it on that list.

>> Also for that Trivy is assuming the wrong package/groupId. The correct groupId is org.eclipse.jdt.core.compiler and Maven central at least has version 4.6.1 : https://repo1.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/

> Can you share link on this jar? Perhaps pom.properties or MANIFEST.MF file contain edu.gmu.cs GroupID.

I already gave you the link : https://repo1.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/ 
The https://repo1.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/4.6.1/ecj-4.6.1.jar does not have any reference to `edu.gmu.cs` GroupID. The `MANIFEST.MF` has `Bundle-SymbolicName: org.eclipse.jdt.core.compiler.batch` and the folders , but the right GAV GroupID is `org.eclipse.jdt.core.compiler`.  What are you looking for in `MANIFEST.MF` to get the `GroupID`? If the header name you look for is `GroupID` then I must say that I have not seen that so far in the jars we use. 

>> Similarly Trivy is assuming the wrong package/groupId also for:
pkg:maven/com.syncthemall/boilerpipe@1.1.0-jahia1
which should be pkg:maven/de.l3s.boilerpipe/boilerpipe@1.1.0-jahia1 [(link to forked artifact)](https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/de/l3s/boilerpipe/boilerpipe/1.1.0-jahia1/) [(link to original on Maven central)](https://repo1.maven.org/maven2/de/l3s/boilerpipe/boilerpipe/1.1.0/)

> New Trivy should not detect this can. Are you using the new Trivy for scanning?

The wrong package/groupID guess was done in Trivy 0.48. The newer Trivy versions are not listing that file at all in the SBOM although it is there. I gave you the link to the forked artifact jar which we use in our image. As this is a forked jar, I hope it is not a big issue to add the `pom.properties` with the GAV.

**Reply by DmitriyLewen** [reply]
> Trivy's is not able to detect the right package name from MANIFEST.MF

Trivy checks `MANIFEST.MF` files. But there is no standard for this file and perhaps we missed something. Can you share examples of files that Trivy doesn't detect - we'll try to update our logic.

---

> Sorry, my mistake. That jar is getting correctly detected. I shouldn't have placed it on that list. 

Do `solr-core@3.6.2-jahia1` and `org.springframework/spring-core@3.2.18.RELEASE_OSGI` work too?


---



> What are you looking for in MANIFEST.MF to get the GroupID?

We take [these fields](https://github.com/aquasecurity/trivy/blob/67c6b1d473999003d682bdb42657bbf3a4a69a9c/pkg/dependency/parser/java/jar/parse.go#L335-L354) from MANIFEST.MF. For GroupID we use [these fields](https://github.com/aquasecurity/trivy/blob/67c6b1d473999003d682bdb42657bbf3a4a69a9c/pkg/dependency/parser/java/jar/parse.go#L391-L408).

---

> I already gave you the link : https://repo1.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/

I realized what the problem is - it is very strange that one organization uses different GroupIDs:
- https://repo1.maven.org/maven2/org/eclipse/jdt/core/maven-metadata.xml
- https://repo1.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/maven-metadata.xml
- https://repo1.maven.org/maven2/org/eclipse/jdt/core/manipulation/maven-metadata.xml

I'll try to update our logic for these cases (we stop checking subdirectories if `metadata.xml` is found, because we thought those directories were version directories).
created https://github.com/aquasecurity/trivy-java-db/issues/30

**Reply by bpapez** [reply]
Thank you for the infos so far. I created an internal ticket to add GAV in `pom.properties` for the libs, which were forked by us or which were never uploaded to Maven central and have no or not enough information in `MANIFEST.MF`.

> Do solr-core@3.6.2-jahia1 and org.springframework/spring-core@3.2.18.RELEASE_OSGI work too?

No they don't work. 

Here is a list of jars, which Trivy >=0.49 does not detect and they do have some information in `MANIFEST.MF` : 

* https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/com/graphql-java-kickstart/graphql-java-kickstart/15.1.1-jahia/graphql-java-kickstart-15.1.1-jahia.jar 
* https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/com/graphql-java-kickstart/graphql-java-servlet/15.1.1-jahia/graphql-java-servlet-15.1.1-jahia.jar 
* https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/apache/solr/solr-core/3.6.2-jahia1/solr-core-3.6.2-jahia1.jar
* https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/apache/solr/solr-solrj/3.6.2-jahia1/solr-solrj-3.6.2-jahia1.jar
* https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/springframework/spring-core/3.2.18.RELEASE_OSGI/spring-core-3.2.18.RELEASE_OSGI.jar
* https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/springframework/spring-expression/3.2.18.RELEASE_OSGI/spring-expression-3.2.18.RELEASE_OSGI.jar

They are all forked by us and are not on Maven central. If you cannot make them detect through given `MANIFEST.MF` info, will it work if we add a GAV in `pom.properties`?

It looks like the next file was even not detected in Trivy 0.48 although it exists on Maven central:
* https://repo1.maven.org/maven2/javax/ccpp/ccpp/1.0/ccpp-1.0.jar

**Reply by DmitriyLewen** [reply]
> will it work if we add a GAV in pom.properties?

yes. Trivy will detect these jars.

> It looks like the next file was even not detected in Trivy 0.48 although it exists on Maven central:
    https://repo1.maven.org/maven2/javax/ccpp/ccpp/1.0/ccpp-1.0.jar

This artifact doesn't have `maven-metadata.xml` file. Therefore, we can't be sure that this file is correct and we don't add similar artifacts in trivy-java-db.


> Here is a list of jars, which Trivy >=0.49 does not detect and they do have some information in MANIFEST.MF :

MANIFEST.MF files don't contain some fields:

- no GroupID: https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/com/graphql-java-kickstart/graphql-java-kickstart/15.1.1-jahia/graphql-java-kickstart-15.1.1-jahia.jar
- no GroupID:    https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/com/graphql-java-kickstart/graphql-java-servlet/15.1.1-jahia/graphql-java-servlet-15.1.1-jahia.jar
- no ArtifactID:    https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/apache/solr/solr-core/3.6.2-jahia1/solr-core-3.6.2-jahia1.jar
- no ArtifactID:    https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/apache/solr/solr-solrj/3.6.2-jahia1/solr-solrj-3.6.2-jahia1.jar
- no GroupID:    https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/springframework/spring-core/3.2.18.RELEASE_OSGI/spring-core-3.2.18.RELEASE_OSGI.jar
- no GroupID:    https://devtools.jahia.com/nexus/content/groups/maven-jahia-org/org/springframework/spring-expression/3.2.18.RELEASE_OSGI/spring-expression-3.2.18.RELEASE_OSGI.jar

**Comment by DmitriyLewen** [comment]
Related issue - https://github.com/aquasecurity/trivy-java-db/issues/28
